File: strain/strain.py
#!/usr/local/bin/python3

'''
This program computes the characteristic strain for
a set of binary black holes for use with Mapelli data
using formulae for the characteristic strain from
Sesana et al., (2005), ApJ, 623, 23.

If the file doesn't exist, the program first creates
a high-resolution array (resolution is defined by the parameter Nz)
containing redshift, lookback-time, and comoving distance
called 'z_tLb_r.dat'

Link to Mapelli paper (arXiv eprint):
https://arxiv.org/pdf/1708.05722.pdf

Link to Sesana paper:
http://iopscience.iop.org/article/10.1086/428492/pdf

Directory structure (user defines BasePath directory):
BasePath/BBHM_DataFiles_Mapelli/
BasePath/strain/
BasePath/strain/hc_DataFiles/
'''

# --- Import Python functions ---
from numpy import pi, sqrt, loadtxt, copy, linspace, where, \
     savetxt, vstack, empty, insert, inf, nan
from sys import exit
from scipy.interpolate import interp1d
from scipy.integrate import romberg
from time import time
import os

ProgStartTime = time()

# --- Define directories ---

# Root directory
#BasePath = '/astro1/dunhamsj/'
BasePath = '/Users/sam/Research/GW/Sam/'

# Directory to strain files
StrainPath = BasePath + 'strain/'

# Input directory for Mapelli data file
InputDir = BasePath + 'BBHM_DataFiles_Mapelli/'

# Input data file name
FileName = 'time_BHillustris1_'

# Output directory for characteristic strain files
OutputDir = StrainPath + 'hc_DataFiles/'

# Header for strain output file
StrainFileHeader = ' hc_{:d}.dat\n\n\
--- Column names ---\n\
(0) Illustris particle ID\n\
(1) Illustris particle initial mass at formation [1e10 Msun/h] (h = 0.704)\n\
(2) Snapshot redshift [dimensionless]\n\
(3) Merger ID\n\
(4) M1 [Msun]\n\
(5) M2 [Msun]\n\
(6) Delay-time [Gyr]\n\
(7) Lookback-time [Gyr]\n\
(8) Redshift at merge [dimensionless]\n\
(9) Comoving distance at merge [Mpc]\n\
(10) fISCO [Hz]\n\
(11) Characteristic strain [dimensionless] ({:d} entries)\n\
(First row contains LISA frequencies)'

# --- All physical quantities are in SI units ---

# --- Physical constants ---
G    = 6.67e-11
c    = 2.99792458e8
year = 86400.0 * 365.25
Gyr  = 1.0e9 * year
Mpc  = 3.086e22
Msun = 2.0e30

# --- Cosmological parameters from Mapelli et al., (2017), MNRAS, 472, 2422
#     (WMAP-9, Hinshaw et al., (2013), ApJS, 208, 19) ---
# NOTE: Hinshaw et al. values do NOT agree with those quoted in Mapelli
h      = 0.704
H0     = h * 100.0 / 3.086e19
OmegaM = 0.2726
OmegaL = 0.7274

# --- Problem-specific constants ---
LISA_Lifetime = 4.0 * year
tau = LISA_Lifetime

# --- Read in LISA data ---
LISA_data = loadtxt( BasePath + 'LISA_sensitivity.dat' )
ff        = LISA_data[:,0]
hc_LISA   = sqrt( LISA_data[:,1] * ff )

# Shorten LISA data
nLISA   = 12
f       = empty( (nLISA), float )
f[0]    = 2.0e-5
f[-1]   = 1.0
f[1:-1] = copy(ff[1::40])

# --- Compute redshift and comoving distance, given lookback time ---
def CreateHighResArray():

    print( 'Creating high-resolution datafile with redshift, lookback-time, \
and comoving distance: z_tLb_r.dat' )

    # Create high-resolution array of redshifts
    Nz = int( 1.0e2 )
    z_arr = linspace( 0.0, 20.0, Nz )

    # Compute lookback-times for redshift array
    def LookbackTimeIntegrand( z ):
        return 1.0 / ( ( 1.0 + z ) * sqrt( OmegaM * ( 1.0 + z )**3 + OmegaL ) )
    def ComputeLookbackTimeFromRedshift( z ):
        return 1.0 / H0 * romberg( LookbackTimeIntegrand, 0.0, z, \
                                     rtol = 1.0e-8 )

    tLb_arr = [ ComputeLookbackTimeFromRedshift( z ) for z in z_arr ]

    # Interpolate redshift array as a function of lookback-time
    z_interp = interp1d( tLb_arr, z_arr, kind = 'linear' )

    # Compute comoving distances for redshift array
    def ComovingDistanceIntegrand( z ):
        return 1.0 / sqrt( OmegaM * ( 1.0 + z )**3 + OmegaL )
    def ComputeComovingDistanceFromRedshift( z ):
        return c / H0 * romberg( ComovingDistanceIntegrand, 0.0, z, \
                                   rtol = 1.0e-8 )
    r_arr = [ ComputeComovingDistanceFromRedshift( z ) for z in z_arr ]


    savetxt( 'z_tLb_r.dat', vstack( ( z_arr, tLb_arr, r_arr ) ), \
               header = 'z_tLb_r.dat; each row has {:d} entries'.format(Nz) )

if not os.path.isfile( StrainPath + 'z_tLb_r.dat' ):
    CreateHighResArray()

# Interpolate lookback-time to get redshift and comoving distance
z_arr, tLb_arr, r_arr = loadtxt( StrainPath + 'z_tLb_r.dat' )
z_interp = interp1d( tLb_arr, z_arr, kind = 'linear', \
                       fill_value = 'extrapolate' )
r_interp = interp1d( tLb_arr, r_arr, kind = 'linear', \
                       fill_value = 'extrapolate' )

# --- Characteristic strain from Sesana et al., (2005), ApJ, 623, 23 ---
def ComputeCharacteristicStrain( M1, M2, tLb ):

    """
    M1, M2 should be in solar masses
    tLb should be in Gyr
    """

    z = z_interp( tLb * Gyr )
    r = r_interp( tLb * Gyr )

    # --- Compute chirp mass (text below Eq. (2)) ---
    Mc = ( M1 * M2 )**( 3.0 / 5.0 ) / ( M1 + M2 )**( 1.0 / 5.0 ) * Msun

    # --- Compute pure strain (Eq. (2)) ---
    h = 8.0 * pi**( 2.0 / 3.0 ) / sqrt( 10.0 ) \
          * ( G * Mc )**( 5.0 / 3.0 ) / ( c**4 * r ) \
          * ( f * ( 1.0 + z ) )**( 2.0 / 3.0 )

    # --- Compute number of cycles in frequency interval around
    #     given frequency f (Eq. (4)) ---
    n = 5.0 / ( 96.0 * pi**( 8.0 / 3.0 ) ) \
          * c**5 / ( G * Mc * f * ( 1.0  + z ) )**( 5.0 / 3.0 )

    # --- Check whether or not n < f * tau and compute characteristic strain
    #     Eq. (6) and Eq. (7) ---
    hc = where( n < f * tau, h * sqrt( n ), h * sqrt( f * tau ) )

    fISCO = c**3 / ( 6.0**(3.0 / 2.0 ) * pi * G ) \
              / ( ( M1 + M2 ) * Msun ) / ( 1.0 + z )

    return z, r, fISCO, hc


# --- Min/max snapshot for which to compute characteristic strains ---
SSmin = 26
SSmax = 135

# --- Number of parameters before strains to print to file ---
nParams = 11

# --- Initialize global min/max values ---
LogFile = StrainPath + 'hc_LogFile.dat'
LogFileFMT = '{:d} {:.16e} {:d} {:.16e} {:.16e} {:.16e} {:.16e} {:.16e} \
{:.16e} {:.16e} {:.16e} {:.16e} {:.16e} {:.16e} {:.16e}\n'

zSnapshotMinG = tDelayMinG = tLbMinG = zMergeMinG = rMinG = +inf
zSnapshotMaxG = tDelayMaxG = tLbMaxG = zMergeMaxG = rMaxG = -inf
TotMergers = 0

with open( LogFile, 'w' ) as fL:
    fL.write( '# Log file\n' )
    fL.write( \
      '# Snapshot, time to read in file [s], nMergers, time to compute hc [s], \
time to save file, zSnapshot Min/Max, tDelay Min/Max [Gyr], tLb Min/Max [Gyr], \
zMerge Min/Max, r Min/Max [Mpc]\n' )

for SS in range( SSmin, SSmax+1 ):
    
    FileIn  = InputDir + FileName + str( SS ) + '.dat'
    FileOut = OutputDir + 'hc_{:d}.dat'.format(SS)

    if not os.path.isfile( FileIn ): continue

    LoadStartTime = time()
    BBHM_Data = loadtxt( FileIn )
    LoadEndTime = time() - LoadStartTime

    IllustrisID      = BBHM_Data[:,0]
    InitialMass      = BBHM_Data[:,1]
    SnapshotRedshift = BBHM_Data[:,2]
    MergerID         = BBHM_Data[:,5]
    M1               = BBHM_Data[:,6]
    M2               = BBHM_Data[:,7]
    tDelay           = BBHM_Data[:,8]
    tLb              = BBHM_Data[:,9]

    N = len(M1)
    TotMergers += N

    hc = empty( ( N+1, nParams + nLISA ), float )
    hc[0,0:nParams] = 'nan'
    hc[0,nParams:]  = f

    StrainStartTime = time()
    for i in range( N ):
        z, r, fISCO, hc[i+1,nParams:] \
          = ComputeCharacteristicStrain( M1[i], M2[i], tLb[i] )
        hc[i+1,0]  = IllustrisID[i]
        hc[i+1,1]  = InitialMass[i]
        hc[i+1,2]  = SnapshotRedshift[i]
        hc[i+1,3]  = MergerID[i]
        hc[i+1,4]  = M1[i]
        hc[i+1,5]  = M2[i]
        hc[i+1,6]  = tDelay[i]
        hc[i+1,7]  = tLb[i]
        hc[i+1,8]  = z
        hc[i+1,9]  = r / Mpc
        hc[i+1,10] = fISCO
    StrainEndTime = time() - StrainStartTime

    SaveStartTime = time()
    savetxt( FileOut, hc, header = StrainFileHeader.format(SS, nLISA) )
    SaveEndTime = time() - SaveStartTime
    LoadEndTime, N, StrainEndTime, SaveEndTime

    # --- Local min/max ---
    zSnapshotMin = tDelayMin = tLbMin = zMergeMin = rMin = +inf
    zSnapshotMax = tDelayMax = tLbMax = zMergeMax = rMax = -inf

    zSnapshotMin = min( zSnapshotMin, hc[i+1,2].min() )
    zSnapshotMax = max( zSnapshotMax, hc[i+1,2].max() )
    tDelayMin    = min( tDelayMin,    hc[i+1,6].min() )
    tDelayMax    = max( tDelayMax,    hc[i+1,6].max() )
    tLbMin       = min( tLbMin,       hc[i+1,7].min() )
    tLbMax       = max( tLbMax,       hc[i+1,7].max() )
    zMergeMin    = min( zMergeMin,    hc[i+1,8].min() )
    zMergeMax    = max( zMergeMax,    hc[i+1,8].max() )
    rMin         = min( rMin,         hc[i+1,9].min() )
    rMax         = max( rMax,         hc[i+1,9].max() )

    with open( LogFile, 'a' ) as fL:
        fL.write( LogFileFMT.format( \
          SS, LoadEndTime, N, StrainEndTime, SaveEndTime, \
          zSnapshotMin, zSnapshotMax, tDelayMin, tDelayMax, tLbMin, \
          tLbMax, zMergeMin, zMergeMax, rMin, rMax ) )

    zSnapshotMinG = min( zSnapshotMin, zSnapshotMinG )
    zSnapshotMaxG = max( zSnapshotMax, zSnapshotMaxG )
    tDelayMinG    = min( tDelayMin,    tDelayMinG    )
    tDelayMaxG    = max( tDelayMax,    tDelayMaxG    )
    tLbMinG       = min( tLbMin,       tLbMinG       )
    tLbMaxG       = max( tLbMax,       tLbMaxG       )
    zMergeMinG    = min( zMergeMin,    zMergeMinG    )
    zMergeMaxG    = max( zMergeMax,    zMergeMaxG    )
    rMinG         = min( rMin,         rMinG         )
    rMaxG         = max( rMax,         rMaxG         )

ProgEndTime = time() - ProgStartTime

with open( LogFile, 'a' ) as fL:
    fL.write( '\n' )
    fL.write( LogFileFMT.format( \
      999, nan, TotMergers, nan, nan, \
      zSnapshotMinG, zSnapshotMaxG, tDelayMinG, tDelayMaxG, tLbMinG, \
      tLbMaxG, zMergeMinG, zMergeMaxG, rMinG, rMaxG ) )
    fL.write( '# Total run-time: {:.16e} s\n'.format(ProgEndTime) )

os.system( 'rm -f *.pyc' )
os.system( 'rm -rf __pycache__/' )

File: strain/strain.f90
PROGRAM CharacteristicStrainFromBBHmergers

  IMPLICIT NONE

  INTEGER,  PARAMETER :: DP = KIND( 1.d0 )
  REAL(DP), PARAMETER :: PI = ACOS( -1.0d0 )

  ! --- Number of quadrature points for look-back time integration ---
  INTEGER,  PARAMETER :: Nq = 10000

  ! --- Resolution of redshift array for reference file ---
  INTEGER, PARAMETER :: Nz = 100000

  ! ---- Misc numbers ---
  INTEGER, PARAMETER :: Nc = 7, Nf_all = 400, Nf = 10, DELTA = Nf_all / Nf

  ! --- Mpc from https://www.cfa.harvard.edu/~dfabricant/huchra/ay145/constants.html ---
  REAL(DP), PARAMETER :: Mpc = 3.0857d24

  ! --- Column labels from Mapelli data ---
  INTEGER, PARAMETER :: iID = 6, iM1 = 7, iM2 = 8, itLB = 10

  ! --- Cosmological parameters come from Mapelli et al., (2017) ----
  REAL(DP), PARAMETER :: OMEGA_M = 0.2726d0, OMEGA_L = 0.7274d0, OMEGA_b = 0.0456d0
  REAL(DP), PARAMETER :: H0 = 70.4d0 * 1.0d5 * ( 1.0d0 / Mpc )

  ! --- Cutoff redshift above which lookback time will be extrapolated ---
  REAL(DP), PARAMETER :: z_max = 25.0d0, dz = z_max / ( Nz - 1 )

  ! --- Physical/unit constants ---
  REAL(DP), PARAMETER :: c = 2.99792458d10, G = 6.673d-8
  REAL(DP), PARAMETER :: Msun = 1.98892d33, Gyr = 1.0d9 * 86400.0d0 * 365.25d0

  ! --- 4 year LISA lifetime ---
  REAL(DP) :: tau = 4.0d0 * 86400.0d0 * 365.25d0

  ! --- Array to hold Illustris data ---
  REAL(DP), ALLOCATABLE :: IllustrisData(:,:)

  ! --- Arrays to hold LISA frequencies ---
  REAL(DP) :: LISA_all(Nf_all,2), LISA(Nf,2)

  ! --- Arrays to hold lookback-time/redshift reference file data ---
  REAL(DP) :: z_arr(Nz), tLb_z_arr(Nz)

  ! --- Misc variables ----
  REAL(DP) :: hc, z, tLb, r, M1, M2, f_ISCO
  INTEGER  :: Nss, SS(2), i, j, k

  ! --- Variables that require more memory ---
  INTEGER( KIND = 8 ) :: iMerger, nLinesIllustris, MergerID

  ! --- For file I/O ---
  INTEGER               :: hcFile = 100, IllustrisDataFile = 101, LISAFile = 102
  INTEGER               :: tLBzFile = 103, NssFile = 104, iArg
  CHARACTER( LEN = 28 ) :: FILEIN
  CHARACTER( LEN = 9 )  :: FMTIN
  CHARACTER( LEN = 3  ) :: arg

  ! --- Boolean for testing code ---
  LOGICAL :: TestStrain_Option = .TRUE.

  ! =================================================================
  !                        Start of program
  ! =================================================================

  ! --- Read in all frequencies from LISA sensitivity curve data file ---
  OPEN( LISAFile, FILE = '../LISA_sensitivity.dat', STATUS = 'OLD' )
  ! --- Loop through comments ---
  DO i = 1, Nc
    READ( LISAFile, * )
  END DO
  ! --- Read in data ---
  DO i = 1, Nf_all
    READ( LISAFile, * ) LISA_all(i,:)
  END DO
  CLOSE( LISAFile )

  ! --- Trim LISA data ---
  LISA(1,:) = LISA_all(1,:) ! --- Make sure first element is kept ---
  DO i = 2, Nf
     LISA(i,:) = LISA_all(i * DELTA,:)
  END DO

  ! --- Create file for storing strains (first row will hold frequencies) ---
  OPEN ( hcFile, FILE = 'hc.dat' )
  WRITE( hcFile, '(A)' ) &
         '# ID, M1, M2, tLb [Gyr], z, f_ISCO, r_comoving [Mpc], hc(f)'
  WRITE( hcFile, &
  '(I7,1x,ES18.10E3,1x,ES18.10E3,1x,ES18.10E3,1x,ES18.10E3,1x,ES18.10E3,1x,ES18.10E3)', &
         ADVANCE = 'NO' ) &
           0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0, 0.0d0
  DO i = 1, Nf
     WRITE( hcFile, '(ES18.10E3)', ADVANCE = 'NO' ) LISA(i,1)
  END DO
  CLOSE( hcFile )

  IF( TestStrain_Option ) THEN

     WRITE(*,'(A)') 'Testing strain calculation...'

     ! --- GW150914 ---
!     M1 = 36.0d0
!     M2 = 29.0d0
!     z  = 0.09d0

     ! --- Sesana et al., (2005), ApJ, 623, 23, Fig. 2 ---
     MergerID = 120
     M1 = 1.0d5
     M2 = 0.1d0 * M1
     z  = 1.0d0
     tau = 3.0d0 * 86400.0d0 * 365.25d0 ! --- Lifetime of three years ---

     CALL main( MergerID, M1, M2, z, ComputeLookBackTime(z) )

     WRITE(*,'(A)') 'Stopping...'
     STOP

  END IF

  ! --- Create lookback time array ---
  OPEN( tLBzFile, FILE = 'tLb_z.dat' )
  WRITE( tLBzFile, '(A)' ) '# z, tLb [Gyr]'
  DO i = 1, Nz
    z_arr(i)     = ( i - 1 ) * dz
    tLb_z_arr(i) = ComputeLookbackTime( z_arr(i) ) / Gyr
    WRITE( tLBzFile, '(ES16.10,1x,F13.10)' ) z_arr(i), tLb_z_arr(i)
  END DO
  CLOSE( tLBzFile )

  ! --- Get upper and lower snapshot numbers ---
  IF ( IARGC() .NE. 2 ) THEN
    WRITE( *, '(A)' ) 'Proper usage: ./strain SSL SSU'
    WRITE( *, '(A)' ) 'Exiting...'
    STOP
  END IF

  ! --- Read upper and lower snapshot numbers into array ---
  DO iArg = 1, IARGC()
    CALL GETARG( iArg, arg )
    READ( arg, * ) SS(iArg)
  END DO

  ! --- Loop through Illustris snapshots ---
  DO Nss = SS(1), SS(2)

    ! --- These two files don't exist ---
    IF ( ( Nss .NE. 53 ) .AND. ( Nss .NE. 55 ) ) THEN

      ! --- Save file with snapshot number currently being analyzed ---    
      OPEN ( NssFile, FILE = 'Nss.dat' )
      WRITE( NssFile, '(I3)' ) Nss
      CLOSE( NssFile )

      ! --- Get filenames ---
      IF ( Nss < 100 ) THEN
        FMTIN = '(A,I2,A4)'
      ELSE
        FMTIN = '(A,I3,A4)'
      END IF
      WRITE( FILEIN, FMTIN  ) &
           '../time_BHillustris1_', Nss, '.dat'
!           '../BBHM_DataFiles_Mapelli/time_BHillustris1_', Nss, '.dat'

      ! --- Get number of lines (mergers) in Illustris data file ---
      nLinesIllustris = 0
      OPEN( IllustrisDataFile, FILE = TRIM( FILEIN ) )
      DO
        READ( IllustrisDataFile, *, END = 11 )
        nLinesIllustris = nLinesIllustris + 1
      END DO
      11 CLOSE( IllustrisDataFile )

      ! --- Read in Illustris data and compute strain for each merger ---
      ALLOCATE( IllustrisData( nLinesIllustris, 10 ) )
      OPEN( IllustrisDataFile, FILE = TRIM( FILEIN ) )
      DO iMerger = 1, nLinesIllustris
        READ( IllustrisDataFile, * ) IllustrisData(iMerger,:)

        ! --- Get data for this merger ---
        MergerID = INT8( IllustrisData(iMerger,iID) )
        M1       = IllustrisData(iMerger,iM1)  ! [ Solar masses, source frame ]
        M2       = IllustrisData(iMerger,iM2)  ! [ Solar masses, source frame ]
        tLb      = IllustrisData(iMerger,itLb) ! [ Gyr ]

        ! --- Compute redshift from lookback time ---
        z = InterpolateLookbackTime( tLb )

        CALL main( MergerID, M1, M2, z, tLb )
        
      END DO ! --- Loop over mergers ---

      CLOSE( IllustrisDataFile )
      DEALLOCATE( IllustrisData )

    END IF ! --- Not using 53 or 55 ---

  END DO ! --- Loop over all snapshots ---


  ! =================================================================
  !                          End of program
  ! =================================================================


CONTAINS

  ! --- Integrand E(z) in cosmological distance calculation ---
  PURE FUNCTION E( z )

    REAL(DP), INTENT(in) :: z
    REAL(DP)             :: E

    E = 1.0d0 / SQRT( ( OMEGA_M + OMEGA_b ) * ( 1.0d0 + z )**3 + OMEGA_L )

    RETURN
  END FUNCTION E

  
  FUNCTION ComputeComovingDistance( z ) RESULT( r )

    REAL(DP), INTENT(in) :: z
    REAL(DP)             :: r, dz

    ! --- Integrate with Trapezoidal rule ---
    r  = 0.0d0
    dz = z / Nq
    
    DO k = 1, Nq - 1
       r = r + E( k * dz )
    END DO

    r = c / H0 * dz / 2.0d0 * ( E( 0.0d0 ) + 2.0d0 * r + E( z ) )

    RETURN
  END FUNCTION ComputeComovingDistance

  
  ! --- Integrand in lookback time calculation ---
  PURE FUNCTION E_LB( z )

    REAL(DP), INTENT(in) :: z
    REAL(DP)             :: E_LB

    E_LB = 1.0d0 / ( ( 1.0d0 + z ) &
             * SQRT( ( OMEGA_M + OMEGA_b ) * ( 1.0d0 + z )**3 + OMEGA_L ) )
    
    RETURN
  END FUNCTION E_LB

  
  FUNCTION ComputeLookbackTime( z ) RESULT( tLB )

    REAL(DP), INTENT(in) :: z
    REAL(DP)             :: dz, tLB
    INTEGER              :: i
    
    ! --- Integrate with Trapezoidal rule ---
    ! --- |ERROR|: Nq = 1000   --> 1.47d-3
    !              Nq = 10000  --> 1.47d-5
    !              Nq = 100000 --> 1.47d-7 ---
    
    tLB = 0.0d0
    dz  = z / Nq

    tLB = 0.0d0
    DO i = 1, Nq - 1
      tLB = tLB + E_LB( i * dz )
    END DO

    tLB = 1.0d0 / H0 * dz / 2.0d0 * ( E_LB( 0.0d0 ) + 2.0d0 * tLB + E_LB( z ) )

    RETURN
  END FUNCTION ComputeLookbackTime

  
  ! --- Characteristic strain (dimensionless)
  !       from Sesana et al. (2005), ApJ, 623, 23: Eqs. (6,7) ---
  FUNCTION ComputeCharacteristicStrain( M1, M2, f, z, r ) RESULT( hc )

    REAL(DP), INTENT(in)  :: M1, M2, f, z, r
    REAL(DP)              :: hc
    REAL(DP)              :: Mc, h, n

    ! --- Compute chirp mass (in source frame) ---
    Mc = ( M1 * M2 )**( 3.0d0 / 5.0d0 ) / ( M1 + M2 )**( 1.0d0 / 5.0d0 ) * Msun

    ! --- Compute pure strain, Eq. (2) ---
    h = 8.0d0 * PI**( 2.0d0 / 3.0d0 ) / SQRT( 10.0d0 )   &
          * ( G * Mc )**( 5.0d0 / 3.0d0 ) / ( c**4 * r ) &
          * ( f * ( 1.0d0 + z ) )**( 2.0d0 / 3.0d0 )

    ! --- Compute n, Eq. (4) ---
    n = 5.0d0 / ( 96.0d0 * PI**( 8.0d0 / 3.0d0 ) ) &
          * c**5 / ( G * Mc * f * ( 1.0d0 + z ) )**( 5.0d0 / 3.0d0 )

    ! --- Check whether or not n < f * tau ---
    IF ( n < f * tau ) THEN
      hc = h * SQRT( n )
    ELSE
      hc = h * SQRT( f * tau )
    END IF
   
    RETURN
  END FUNCTION ComputeCharacteristicStrain

  
  ! --- Interpolate lookback time array to get redshift ---
  FUNCTION InterpolateLookbackTime( tLB ) RESULT( z )

    ! --- Get redshift by interpolating lookback time ---
    ! --- tLB should be in units of Gyr ---

    REAL(DP), INTENT(in) :: tLB
    REAL(DP)             :: zmin, zmax, tLBmin, tLBmax, tLBi, m, b
    REAL(DP)             :: z
    INTEGER              :: i

    ! --- Small lookback time approximation: tLB ~ tH * z ---
    ! --- 0.1 chosen to ensure 1% error in distances to match
    !       LISA's predicted capability (Ask Kelly for reference to this) ---
    IF ( tLB < 0.1d0 ) THEN
      z = tLB / ( ( 1.0d0 / H0 ) / Gyr )
      RETURN
    END IF

    ! --- If target time is in the data set,
    !       interpolate using linear interpolation ---
    IF ( tLB < tLB_z_arr(Nz) ) THEN

      ! --- Get redshift bounds ---
      i = 1
      tLBi = tLB_z_arr(i)
      DO WHILE ( tLBi < tLB )
        i = i + 1
        tLBi = tLB_z_arr(i)
      END DO

      zmin   = z_arr(i-1)
      zmax   = z_arr(i)
      tLBmin = tLB_z_arr(i-1)
      tLBmax = tLB_z_arr(i)

    ! --- If target time is not in the data set,
    !       extrapolate using linear extrapolation ---
    ELSE

      zmin   = z_arr(Nz-1)
      zmax   = z_arr(Nz)
      tLBmin = tLB_z_arr(Nz-1)
      tLBmax = tLB_z_arr(Nz)

    END IF

    ! --- tLB = m * z + b ---
    m = ( tLBmax - tLBmin ) / ( zmax - zmin )
    b = 0.5_DP * ( ( tLBmax - m * zmax ) + ( tLBmin - m * zmin ) )

    z = ( tLB - b ) / m

    RETURN
  END FUNCTION InterpolateLookbackTime


  SUBROUTINE main( MergerID, M1, M2, z, tLb )

    INTEGER( KIND = 8 ), INTENT(in) :: MergerID
    REAL(DP),            INTENT(in) :: M1, M2, z, tLb

    ! --- Compute comoving distance (in cm) ---
    r = ComputeComovingDistance( z ) ! [ cm ]

    ! --- Calculate frequency at ISCO using
    !       Sesana et al. (2005), Eq. (12) ---
    f_ISCO = c**3 / ( 6.0d0**( 3.0d0 / 2.0d0 ) * PI * G &
               * ( M1 + M2 ) * Msun  * ( 1.0d0 + z ) )

    ! --- Start writing to file ---
    OPEN ( hcFile, FILE = 'hc.dat', POSITION = 'APPEND' )
    WRITE( hcFile, &
   '(I7,1x,ES18.10E3,1x,ES18.10E3,1x,ES18.10E3,1x,ES18.10E3,1x,ES18.10E3,1x,ES18.10E3)', &
             ADVANCE = 'NO' ) MergerID, M1, M2, tLB, z, f_ISCO, r / Mpc
        
    ! --- Loop through frequencies until f_ISCO and write to file ---
    j = 1
    DO WHILE ( LISA(j,1) .LT. f_ISCO )
      hc = ComputeCharacteristicStrain( M1, M2, LISA(j,1) ,z ,r )
      WRITE( hcFile, '(ES13.6)', ADVANCE = 'NO' ) hc
      j = j + 1
      IF( j .GT. Nf ) EXIT
    END DO

    ! --- Fill in missing frequencies with nan and write to file ---
    DO WHILE ( j .LT. Nf + 1 )
      WRITE( hcFile, '(A4)', ADVANCE = 'NO' ) ' nan'
      j = j + 1
    END DO

    WRITE( hcFile, * )
    CLOSE( hcFile )

    RETURN
  END SUBROUTINE main

END PROGRAM CharacteristicStrainFromBBHmergers
